File: miscellaneous/tree-bfs/20221222/tree_bfs.py
from __future__ import annotations
from typing import List


def bfs(tree: TreeNode) -> List[int]:
    queue: List[TreeNode] = [tree]
    values: List[int] = []

    while len(queue):
        node = queue.pop(0)
        values.append(node.value)
        node.left and queue.append(node.left)
        node.right and queue.append(node.right)

    return values


class TreeNode:
    value: int
    left: TreeNode
    right: TreeNode

    def __init__(self, value: int, left:TreeNode=None, right:TreeNode=None):
        self.value = value
        self.left = left
        self.right = right

File: miscellaneous/tree-bfs/20221222/tree-bfs.js
'use strict';

exports.treeBfs = (head) => {
    const values = [];
    const queue = [head];
    const seen = new Set();
    seen.add(head.val);

    while (queue.length) {
        const node = queue.shift();
        values.push(node.val);
        node.left && !seen.has(node.left) && queue.push(node.left);
        node.right && !seen.has(node.right) && queue.push(node.right);
    }

    return values;
}

exports.TreeNode = class {
    constructor(val, left=null, right=null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}